scintillometry/fourier/base.py

```diff
@@ -0,0 +1,273 @@
+            if real_transform:
+                output_shape[axis] = output_shape[axis] * 2
+                output_dtype = np.dtype('float{0:d}'.format(
+                    input_data['dtype'].itemsize * 8 // 2))
+            else:
+                output_dtype = input_data['dtype']
+
+            freq_data = input_data
+            time_data = {'shape': tuple(output_shape),
+                         'dtype': output_dtype}
+        else:
+            if 'float' in input_data['dtype'].name:
+                # assert not (input_data['shape'][axis] % 2), (
+                #     "time-domain array must have an even number of elements "
+                #     "along the axis being transformed.")
+                output_shape[axis] = output_shape[axis] // 2 + 1
+                output_dtype = np.dtype('complex{0:d}'.format(
+                    input_data['dtype'].itemsize * 8 * 2))
+            else:
+                output_dtype = input_data['dtype']
+
+            time_data = input_data
+            freq_data = {'shape': tuple(output_shape),
+                         'dtype': output_dtype}
+
+        # Store time and frequency-domain array shapes.
+        data_format = {'time_shape': time_data['shape'],
+                       'time_dtype': time_data['dtype'],
+                       'freq_shape': freq_data['shape'],
+                       'freq_dtype': freq_data['dtype']}
+
+        return self._setup_transform(data_format, direction, axis, norm,
+                                     sample_rate)
+
+    def _setup_transform(data_format, direction, axis, norm, sample_rate):
+        raise NotImplementedError()
+
+    def inverse(self, data_format, direction, axis, norm, sample_rate):
+        inverse_direction = 'forward' if direction == 'inverse' else 'inverse'
+        return self._setup_transform(data_format, inverse_direction,
+                                     axis, norm, sample_rate)
+
+
+class FFT(object):
+    """Single pre-defined FFT and its associated metadata."""
+
+    def __init__(self, fft, data_format, direction, axis, norm,
+                 sample_rate, parent):
+        self._fft = fft
+        self._data_format = data_format
+        self._direction = direction
+        self._axis = axis
+        self._norm = norm
+        self._sample_rate = sample_rate
+        self._parent = parent
+
+    @property
+    def data_format(self):
+        """Shapes and dtypes of the FFT arrays.
+
+        'time_' and 'freq_' entries are for time and frequency-domain arrays,
+        respectively.
+        """
+        return self._data_format
+
+    @property
+    def direction(self):
+        """Axis over which to perform the FFT."""
+        return self._direction
+
+    @property
+    def axis(self):
+        """Axis over which to perform the FFT."""
+        return self._axis
+
+    @property
+    def norm(self):
+        """Normalization convention.
+
+        As in `numpy.fft`, `None` is an unscaled forward transform and 1 / n
```

Comment:
Here, I think it should just be defined (i.e., translate if `self._norm is None`)